### DISCUSSION TITLE
Unhelpful libtool error when compiling on Centos 7

### DISCUSSION THREAD

**Discussion by nneil** [ImageMagick/ImageMagick#6682] [Contribute]
### ImageMagick version

7.1.0-62

### Operating system

Linux

### Operating system, version and so on

CentOS Linux release 7.9.2009 (Core)

### Description

Unhelpful compile time error. 

gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
GNU C Library (GNU libc) stable release version 2.17, by Roland McGrath et al.

```
  CC       MagickWand/libMagickWand_7_Q16HDRI_la-wand.lo
  CC       MagickWand/libMagickWand_7_Q16HDRI_la-wandcli.lo
  CC       MagickWand/libMagickWand_7_Q16HDRI_la-wand-view.lo
  CCLD     MagickWand/libMagickWand-7.Q16HDRI.la
  CXXLD    utilities/magick
libtool:   error: unsupported hardcode properties See the libtool documentation for more information. Fatal configuration error.
make[1]: *** [utilities/magick] Error 1
make[1]: Leaving directory `/home/dam/build/ImageMagick-7.1.0-62'
make: *** [all] Error 2
```

### Steps to Reproduce

```
./configure
make
```

**Comment by urban-warrior** [comment]
We didn't write nor maintain libtool.  Consider their advice and see the libtool document for more information.

**Comment by nneil** [comment]
You didn't write it, but you are using it. Is there any chance that you might be using it wrongly?

Centos was at one time very popular and 7.0 is still supported. It wouldn't be unreasonable for you to have compiled on Centos and learnt something from that process. You can't expect people who simply use your software to be as familiar with libtool as you are.

"Hardcode" appears in limited parts of the libtool documentation, which not aimed at typical sysadmins. I didn't find it helpful.

**Comment by dlemstra** [comment]
We suspect that this is something that is related to your operating system or compiler tools that you are using. You can't expect us to setup a Centos environment that uses a compiler that is no longer supported to investigate this for you.

**Comment by nneil** [comment]
I was simply hoping for a clue as to how to proceed. I'm not happy supporting old systems, but it's where I find myself.

It looks like the problem was simply that there was no C++ compile installed, only C. Having looked through the configure output I installed the C++ compiler and it now compiles.

But shouldn't this be detected by ./configure? I am used to the configure stage detecting missing libraries, etc.

**Comment by fmw42** [comment]
Look through config.log.  It should show any errors.  Always check the config.log for error and clues where to look.

**Comment by codereverser** [comment]
Can you try installing cxx (`g++`) compiler and see if it resolves the issue? 

Looks like https://github.com/ImageMagick/ImageMagick/commit/d6061d9bbc646bf7f7ef467b0d375cb44064a890#diff-b48650cabc0f502b52865f525765f2d688198f06e0ffb5456c76fabab7f2bfb2R2132 switched the linker for `utilities/magick` to cxx instead of C and libtool fails during the link stage (at least in my case)

**Comment by nneil** [comment]
configure did not detect the absence of a c++ compiler. Once I installed one I was able to compile the project.

I request that the configure script warn if necessary packages are not installed on the system.

**Comment by AidenMic** [comment]
Edit: Wait

I'm running a server on Ubuntu 22.04 and I'm getting the same (or at least a very similar) error

**ImageMagick Version**
7.1.1-16
**Operating System Etc**
Ubuntu 22.043 LTS (Jammy Jellyfish)
gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0

I looked at the config log, makefile, and piped the ./configure and make (and gmake for good measure), but this is my first time installing ImageMagick and I'm not sure what I should be looking for. I uploaded these files with my username and server IP removed if anyone with more experience wants to take a look. In the meantime I'll try look at the Libtool documentation, although I'm not sure where to start
[config.log](https://github.com/ImageMagick/ImageMagick/files/12449145/config.log)
[ConfigureOutputLog.txt](https://github.com/ImageMagick/ImageMagick/files/12449146/ConfigureOutputLog.txt)
[MakeOutputLog.txt](https://github.com/ImageMagick/ImageMagick/files/12449144/MakeOutputLog.txt)
[GMakeOutputLog.txt](https://github.com/ImageMagick/ImageMagick/files/12449143/GMakeOutputLog.txt)
[Makefile.txt](https://github.com/ImageMagick/ImageMagick/files/12449168/Makefile.txt)

**Reply by urban-warrior** [reply]
It's looking for a usable C++ compiler.  Install the c++ compiler or add `--without-magick-plus-plus` to your configure command-line.  Does ImageMagick build now?

**Reply by Tors-0** [reply]
had this problem on linux mint 21.3, had to install `libc6-dev` and `g++` and then rerun ./configure and make and now it works

**Comment by urban-warrior** [comment]
Try

```
sudo apt install libtool-dev
```
Try make again.  Does it work now?

**Comment by AidenMic** [comment]
> Try
> 
> ```
> sudo apt install libtool-dev
> ```
> 
> Try make again. Does it work now?

It just returns `E: Unable to locate package libtool-dev` 
I did realize I didn't have the libtool-bin installed literally just after my second edit (after I tried to check my version like the libtool doc suggests), but I was able to install that just fine and `make` still failed. Is libtool-dev the right package name?

**Comment by urban-warrior** [comment]
Try `apt install libltdl-dev`.

**Comment by AidenMic** [comment]
It says that `libltdl-dev is already the newest version (2.4.6-15build2).` That said I'll try rebooting my server really quick.

edit: OK, I rebooted and it's still giving the same error

**Comment by fheyer** [comment]
I had exactly this issue on Ubuntu 20.04
* build stopped with message

    CXXLD    utilities/magick
    libtool:   error: unsupported hardcode properties See the libtool documentation for more information. Fatal configuration error.

Examination showed i had no C++ compiler installed. After installing one with `apt-get install g++` and after cleaning up with `make clean` compilation went through without error.

**Comment by MoonCactus** [comment]
Tried the various suggestions made here, but none fixed the issue for me - Debian 6.5.13-1 (2023-11-29) x86_64 GNU/Linux
Using `./configure --disable-dependency-tracking` fixed it.
I did not check any possible side-effects though so YMMV.

**Reply by Danack** [reply]
This appears to have fixed it for me. The other suggestions didn't appear to have any effect.

**Comment by ll0224** [comment]
I encountered libtool: error: unsupported hardcode properties See the libtool documentation for more information during the compilation and installation of ImageMagick-7.1.1-26 Fatal configuration error
Install gcc-c++, `yum install gcc-c++`, and execute again `.configure` final `make`

**Comment by LeftMonster** [comment]
I have been trying to build my own Dockerfile, but I have been failing for the past two days.
I then tried installing additional libraries, which were displayed in the build logs.
Maybe everyone is different!

Got it!~😄
终于搞定！！~😄

Here is my machine information:
> Ubuntu 20.04

My solution:
```bash
# Add g++ and use ./configure --disable-dependency-tracking
# For more details, you can refer to the Dockerfile code block below
# part one
apt install g++
# part two
./configure --disable-dependency-tracking
```

```dockerfile
RUN apt-get update && \
    apt-get install -y \
        gcc \
        g++ \
        wget \
        make \
        imagemagick && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

RUN wget -c https://github.com/ImageMagick/ImageMagick/archive/refs/tags/7.1.1-35.tar.gz -O - | tar -xz && \
    cd ImageMagick-7.1.1-35 && \
    ./configure --disable-dependency-tracking && \
    make && make install && \
    ldconfig /usr/local/lib && \
    cd .. && rm -rf ImageMagick-7.1.1-35/
```
